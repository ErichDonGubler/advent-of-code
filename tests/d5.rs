use std::collections::{HashMap, HashSet};

use itertools::Itertools;

const EXAMPLE: &str = "\
47|53
97|13
97|61
97|47
75|29
61|13
75|53
29|13
97|29
53|29
61|53
97|53
61|29
47|13
75|47
97|75
47|61
75|61
47|29
75|13
53|13

75,47,61,53,29
97,61,53,29,13
75,29,13
75,97,47,61,53
61,13,29
97,13,75,29,47
";

pub(crate) struct BeforeAfterRules {
    inner: HashMap<u8, HashSet<u8>>,
}

impl BeforeAfterRules {
    pub fn breaks_with(&self, before: u8, after: u8) -> bool {
        let Self { inner } = self;
        inner
            .get(&after)
            .is_some_and(|needs_after| needs_after.contains(&before))
    }
}

fn parse_p1(input: &str) -> (BeforeAfterRules, impl Iterator<Item = Vec<u8>> + '_) {
    let mut lines = input.lines();

    let mut before_after_rules = HashMap::<_, HashSet<_>>::new();
    lines
        .by_ref()
        .take_while(|line| !line.trim().is_empty())
        .for_each(|line| {
            let (before, after) = line
                .splitn(2, "|")
                .map(|x| x.parse().unwrap())
                .collect_tuple()
                .expect("expected rule with `|`");
            before_after_rules.entry(before).or_default().insert(after);
        });

    let updates = lines.map(|l| {
        l.split(",")
            .map(|x| x.parse().unwrap())
            .collect::<Vec<u8>>()
    });

    (
        BeforeAfterRules {
            inner: before_after_rules,
        },
        updates,
    )
}

fn correct_middle_page_number_sum(input: &str) -> u32 {
    let (before_after_rules, updates) = parse_p1(input);

    let mut middle_page_number_sum = 0u32;
    'next_row: for update in updates {
        let mut update_iter = update.iter().copied();
        while let Some(value) = update_iter.next() {
            for other_value in update_iter.clone() {
                if before_after_rules.breaks_with(value, other_value) {
                    continue 'next_row;
                }
            }
        }
        // TODO: Is this indexing right? The instructions say nothing about
        // even-numbered update rows. ðŸ«¤
        middle_page_number_sum = middle_page_number_sum
            .checked_add(update[update.len() / 2].into())
            .unwrap();
    }
    middle_page_number_sum
}

#[test]
fn p1_example() {
    assert_eq!(correct_middle_page_number_sum(EXAMPLE), 143);
}

const INPUT: &str = include_str!("./d5.txt");

#[test]
fn p1() {
    assert_eq!(correct_middle_page_number_sum(INPUT), 7024);
}

fn incorrect_middle_page_number_sum(input: &str) -> u32 {
    let (before_after_rules, updates) = parse_p1(input);

    let mut incorrect_middle_page_number_sum = 0u32;
    for mut update in updates {
        let mut needed_correction = false;
        let mut value_idx = 0;
        'outer: while value_idx < update.len() {
            let mut other_value_idx = value_idx + 1;
            while other_value_idx < update.len() {
                let other_value = update[other_value_idx];
                if before_after_rules.breaks_with(update[value_idx], other_value) {
                    needed_correction = true;
                    update.remove(other_value_idx);
                    update.insert(value_idx, other_value);
                    continue 'outer;
                }
                other_value_idx += 1;
            }
            value_idx += 1;
        }
        // TODO: Is this indexing right? The instructions say nothing about
        // even-numbered update rows. ðŸ«¤
        if needed_correction {
            incorrect_middle_page_number_sum = incorrect_middle_page_number_sum
                .checked_add(update[update.len() / 2].into())
                .unwrap();
        }
    }
    incorrect_middle_page_number_sum
}

#[test]
fn p2_example() {
    assert_eq!(incorrect_middle_page_number_sum(EXAMPLE), 123);
}

#[test]
fn p2() {
    assert_eq!(incorrect_middle_page_number_sum(INPUT), 4151);
}
