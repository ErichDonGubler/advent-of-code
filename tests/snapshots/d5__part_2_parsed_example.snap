---
source: tests/d5.rs
expression: example_almanac_config
---
AlmanacConfig {
    seeds: [
        Id {
            value: RawId(
                79,
            ),
            space: Seed,
        },
        Id {
            value: RawId(
                80,
            ),
            space: Seed,
        },
        Id {
            value: RawId(
                81,
            ),
            space: Seed,
        },
        Id {
            value: RawId(
                82,
            ),
            space: Seed,
        },
        Id {
            value: RawId(
                83,
            ),
            space: Seed,
        },
        Id {
            value: RawId(
                84,
            ),
            space: Seed,
        },
        Id {
            value: RawId(
                85,
            ),
            space: Seed,
        },
        Id {
            value: RawId(
                86,
            ),
            space: Seed,
        },
        Id {
            value: RawId(
                87,
            ),
            space: Seed,
        },
        Id {
            value: RawId(
                88,
            ),
            space: Seed,
        },
        Id {
            value: RawId(
                89,
            ),
            space: Seed,
        },
        Id {
            value: RawId(
                90,
            ),
            space: Seed,
        },
        Id {
            value: RawId(
                91,
            ),
            space: Seed,
        },
        Id {
            value: RawId(
                92,
            ),
            space: Seed,
        },
        Id {
            value: RawId(
                55,
            ),
            space: Seed,
        },
        Id {
            value: RawId(
                56,
            ),
            space: Seed,
        },
        Id {
            value: RawId(
                57,
            ),
            space: Seed,
        },
        Id {
            value: RawId(
                58,
            ),
            space: Seed,
        },
        Id {
            value: RawId(
                59,
            ),
            space: Seed,
        },
        Id {
            value: RawId(
                60,
            ),
            space: Seed,
        },
        Id {
            value: RawId(
                61,
            ),
            space: Seed,
        },
        Id {
            value: RawId(
                62,
            ),
            space: Seed,
        },
        Id {
            value: RawId(
                63,
            ),
            space: Seed,
        },
        Id {
            value: RawId(
                64,
            ),
            space: Seed,
        },
        Id {
            value: RawId(
                65,
            ),
            space: Seed,
        },
        Id {
            value: RawId(
                66,
            ),
            space: Seed,
        },
        Id {
            value: RawId(
                67,
            ),
            space: Seed,
        },
    ],
    maps: [
        Map {
            from_space: "seed",
            to_space: "soil",
            entries: [
                MapEntry {
                    start_id_from: RawId(
                        50,
                    ),
                    start_id_to: RawId(
                        52,
                    ),
                    size: 48,
                },
                MapEntry {
                    start_id_from: RawId(
                        98,
                    ),
                    start_id_to: RawId(
                        50,
                    ),
                    size: 2,
                },
            ],
        },
        Map {
            from_space: "soil",
            to_space: "fertilizer",
            entries: [
                MapEntry {
                    start_id_from: RawId(
                        0,
                    ),
                    start_id_to: RawId(
                        39,
                    ),
                    size: 15,
                },
                MapEntry {
                    start_id_from: RawId(
                        15,
                    ),
                    start_id_to: RawId(
                        0,
                    ),
                    size: 37,
                },
                MapEntry {
                    start_id_from: RawId(
                        52,
                    ),
                    start_id_to: RawId(
                        37,
                    ),
                    size: 2,
                },
            ],
        },
        Map {
            from_space: "fertilizer",
            to_space: "water",
            entries: [
                MapEntry {
                    start_id_from: RawId(
                        0,
                    ),
                    start_id_to: RawId(
                        42,
                    ),
                    size: 7,
                },
                MapEntry {
                    start_id_from: RawId(
                        7,
                    ),
                    start_id_to: RawId(
                        57,
                    ),
                    size: 4,
                },
                MapEntry {
                    start_id_from: RawId(
                        11,
                    ),
                    start_id_to: RawId(
                        0,
                    ),
                    size: 42,
                },
                MapEntry {
                    start_id_from: RawId(
                        53,
                    ),
                    start_id_to: RawId(
                        49,
                    ),
                    size: 8,
                },
            ],
        },
        Map {
            from_space: "water",
            to_space: "light",
            entries: [
                MapEntry {
                    start_id_from: RawId(
                        18,
                    ),
                    start_id_to: RawId(
                        88,
                    ),
                    size: 7,
                },
                MapEntry {
                    start_id_from: RawId(
                        25,
                    ),
                    start_id_to: RawId(
                        18,
                    ),
                    size: 70,
                },
            ],
        },
        Map {
            from_space: "light",
            to_space: "temperature",
            entries: [
                MapEntry {
                    start_id_from: RawId(
                        45,
                    ),
                    start_id_to: RawId(
                        81,
                    ),
                    size: 19,
                },
                MapEntry {
                    start_id_from: RawId(
                        64,
                    ),
                    start_id_to: RawId(
                        68,
                    ),
                    size: 13,
                },
                MapEntry {
                    start_id_from: RawId(
                        77,
                    ),
                    start_id_to: RawId(
                        45,
                    ),
                    size: 23,
                },
            ],
        },
        Map {
            from_space: "temperature",
            to_space: "humidity",
            entries: [
                MapEntry {
                    start_id_from: RawId(
                        0,
                    ),
                    start_id_to: RawId(
                        1,
                    ),
                    size: 69,
                },
                MapEntry {
                    start_id_from: RawId(
                        69,
                    ),
                    start_id_to: RawId(
                        0,
                    ),
                    size: 1,
                },
            ],
        },
        Map {
            from_space: "humidity",
            to_space: "location",
            entries: [
                MapEntry {
                    start_id_from: RawId(
                        56,
                    ),
                    start_id_to: RawId(
                        60,
                    ),
                    size: 37,
                },
                MapEntry {
                    start_id_from: RawId(
                        93,
                    ),
                    start_id_to: RawId(
                        56,
                    ),
                    size: 4,
                },
            ],
        },
    ],
}
